### DISCUSSION TITLE
MERGE support for HQL/Criteria

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#5896] [Design Proposals]
Support the SQL standard `MERGE` statement. Essentially it's like `INSERT` with the `CONFLICT` clause, but also offers the possibility to do `DELETE`s/`UPDATE`s. Quite a few DBMS support this. The concurrency guarantees of the statement AFAIU allows such a statement to be implemented by executing the individual parts as separate statements, so emulation should be possible.

Also see:

* https://en.wikibooks.org/wiki/Structured_Query_Language/MERGE
* https://en.wikipedia.org/wiki/Merge_(SQL)

Syntax-wise I would suggest we follow the SQL standard approach which seems sane to me.

```antlr
merge_statement : MERGE INTO entityName (AS identifier)? USING fromRoot ON predicate merge_when_clause+;
merge_when_clause
    : WHEN MATCHED (AND predicate)? THEN (UPDATE setClause | DELETE)
    | WHEN NOT MATCHED (AND predicate)? THEN insertStatement
    ;
```

**Comment by dreab8** [comment]
https://hibernate.atlassian.net/browse/HHH-19404
